File: 2015/16/rust/src/main.rs
use regex::Regex;
use std::collections::HashMap;
use std::fs::File;
use std::io::{ BufRead, BufReader };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = r"([a-z]+): (\d+)";

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let regex = Regex::new(PARSER_REGEX)?;

    // --- MFCSAM Output ---
    let mut map: HashMap<String, usize> = HashMap::new();
    map.insert("children".to_string(), 3);
    map.insert("cats".to_string(), 7);
    map.insert("samoyeds".to_string(), 2);
    map.insert("pomeranians".to_string(), 3);
    map.insert("akitas".to_string(), 0);
    map.insert("vizslas".to_string(), 0);
    map.insert("goldfish".to_string(), 5);
    map.insert("trees".to_string(), 3);
    map.insert("cars".to_string(), 2);
    map.insert("perfumes".to_string(), 1);

    // --- Read and parse the input file ---
    let file = File::open(INPUT_FILE)?;
    'outer: for line in BufReader::new(file).lines() {
        let line = line.unwrap();

        // --- Puzzle 1 ---
        for cap in regex.captures_iter(&line) {
            let key = cap[1].to_string();
            let value = cap[2].parse::<usize>()?;

            if let Some(v) = map.get(&key) {
                if *v != value { continue 'outer; }
            }
        }
        let sue = line.split(":").next().unwrap();
        println!("1. Number of the Sue that got you the gift: {}", sue);
        break;
    }

    let file = File::open(INPUT_FILE)?;
    'outer: for line in BufReader::new(file).lines() {
        let line = line.unwrap();

        // --- Puzzle 2 ---
        for cap in regex.captures_iter(&line) {
            let key = cap[1].to_string();
            let value = cap[2].parse::<usize>()?;

            if let Some(v) = map.get(&key) {
                if key == "cats" || key == "trees" {
                    if *v >= value { continue 'outer; }
                } else if key == "pomeranians" || key == "goldfish" {
                    if *v <= value { continue 'outer; }
                } else {
                    if *v != value { continue 'outer; }
                }
            }
        }
        let sue = line.split(":").next().unwrap();
        println!("2. Number of the Sue that got you the gift: {}", sue);
        break;
    }

    Ok(())
}

File: 2015/16/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"
const REGEX = `([a-z]+): (\d+)`

// --- MFCSAM Output ---
var MAP = map[string]int{
	"children":    3,
	"cats":        7,
	"samoyeds":    2,
	"pomeranians": 3,
	"akitas":      0,
	"vizslas":     0,
	"goldfish":    5,
	"trees":       3,
	"cars":        2,
	"perfumes":    1,
}

// --- Functions ---
func find1(line string, re *regexp.Regexp) bool {
	ms := re.FindAllStringSubmatch(line, -1)
	for _, m := range ms {
		key := m[1]
		val, err := strconv.Atoi(m[2])
		if err != nil {
			log.Panic(err)
		}
		if v, ok := MAP[key]; ok && v != val {
			return false
		}
	}
	return true
}

func find2(line string, re *regexp.Regexp) bool {
	ms := re.FindAllStringSubmatch(line, -1)
	for _, m := range ms {
		key := m[1]
		val, err := strconv.Atoi(m[2])
		if err != nil {
			log.Panic(err)
		}
		if v, ok := MAP[key]; ok {
			switch key {
			case "cats", "trees":
				if v >= val {
					return false
				}
			case "pomeranians", "goldfish":
				if v <= val {
					return false
				}
			default:
				if v != val {
					return false
				}
			}
		}
	}
	return true
}

func main() {
	// --- Variables ---
	re := regexp.MustCompile(REGEX)
	sue1 := ""
	sue2 := ""

	// --- Read input file ---
	file, err := os.Open(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		// --- Puzzle 1 ---
		if sue1 == "" && find1(line, re) {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) > 0 {
				sue1 = strings.TrimSpace(parts[0])
			}
		}

		// --- Puzzle 2 ---
		if sue2 == "" && find2(line, re) {
			parts := strings.SplitN(line, ":", 2)
			if len(parts) > 0 {
				sue2 = strings.TrimSpace(parts[0])
			}
		}
	}

	// --- Results ---
	fmt.Println("1. Number of the Sue that got you the gift:", sue1)
	fmt.Println("2. Number of the Sue that got you the gift:", sue2)
}
